python-package/lightgbm/basic.py

```diff
@@ -1291,6 +1518,75 @@ def _lazy_init(self, data, label=None, reference=None,
         # set feature names
         return self.set_feature_name(feature_name)
 
+    def __yield_row_from(self, seqs: List[Sequence], indices: Iterable[int]):
+        offset = 0
+        seq_id = 0
+        seq = seqs[seq_id]
+        for row_id in indices:
+            assert row_id >= offset, "sample indices are expected to be monotonic"
+            while row_id >= offset + len(seq):
+                offset += len(seq)
+                seq_id += 1
+                seq = seqs[seq_id]
+            id_in_seq = row_id - offset
+            row = seq[int(id_in_seq)]
+            yield row if row.flags['OWNDATA'] else row.copy()
+
+    def __sample(self, seqs: List[Sequence], total_nrow: int) -> Tuple[List[np.ndarray], List[np.ndarray]]:
+        """Sample data from seqs.
+
+        Mimics behavior in c_api.cpp:LGBM_DatasetCreateFromMats()
+
+        Returns
+        -------
+            sampled_rows, sampled_row_indices
+        """
+        indices = self.create_sample_indices(total_nrow)
+
+        # Select sampled rows, transpose to column order.
+        sampled = np.array([row for row in self.__yield_row_from(seqs, indices)])
+        sampled = sampled.T
+
+        filtered = []
+        filtered_idx = []
+        sampled_row_range = np.arange(len(indices), dtype=np.int32)
+        for col in sampled:
+            col_predicate = (np.abs(col) > ZERO_THRESHOLD) | np.isnan(col)
+            filtered_col = col[col_predicate]
+            filtered_row_idx = sampled_row_range[col_predicate]
+
+            filtered.append(filtered_col)
+            filtered_idx.append(filtered_row_idx)
+
+        return filtered, filtered_idx
+
+    def __init_from_seqs(self, seqs: List[Sequence], ref_dataset: 'Dataset'):
+        """
+        Initialize data from list of Sequence objects.
+
+        Sequence: Generic Data Access Object
+            Supports random access and access by batch if properly defined by user
+
+        Data scheme uniformity are trusted, not checked
+        """
+        total_nrow = sum(len(seq) for seq in seqs)
+
+        # create validation dataset from ref_dataset
+        if ref_dataset:
```

Comment:
I believe more explicit check is better.
```suggestion
        if ref_dataset is not None:
```